File: TTT-Library/Judge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TTT_Library
{
    public class Judge
    {
        public char Winner { get; private set; }

        private AbstractGameBoard _board;
        private readonly List<AbstractTile> _currentLine = new();
        private bool _winnerFound = false;

        public Judge(AbstractGameBoard gameboard)
        {
            Winner = 'N';
            _board = gameboard;
        }

        public bool FindsWinner()
        {
            if (!_winnerFound)
            {
                CheckForWinner();
            }
            return _winnerFound;
        }

        private void CheckForWinner()
        {
            for (int i = 0; i < 3; i++)
            {
                for (int j = 0; j < 3; j++)
                {
                    Explore(i, j, new Impulse(0, 1));
                    Explore(i, j, new Impulse(1, 0));
                    Explore(i, j, new Impulse(1, 1));
                    Explore(i, j, new Impulse(1, -1));
                }
            }
        }

        private bool ValidCoordinates(int i, int j)
        {
            return (0 <= i && i < 3) && (0 <= j && j < 3);
        }

        private void Explore(int i, int j, Impulse impulse)
        {
            if (ValidCoordinates(i, j))
            {
                AbstractTile tile = _board.GetTile(i, j);
                _currentLine.Add(tile);

                if(_currentLine.Count < 3)
                {
                    impulse.Translate(ref i, ref j);
                    Explore(i, j, impulse);
                }
                else
                {
                    CheckCurrentLine();
                }

                _currentLine.Remove(tile);
            }
        }

        private void CheckCurrentLine()
        {
            if (IsWinningLine(_currentLine))
            {
                _winnerFound = true;
                Winner = _currentLine[0].Character;
            }
        }

        private bool IsWinningLine(List<AbstractTile> line)
        {
            Dictionary<char, int> table = new()
            {
                { ' ', 0 },
                { 'X', 0 },
                { 'O', 0 },
            };

            foreach (AbstractTile tile in line)
            {
                table[tile.Character] += 1;
            }

            return (table['X'] == 3 || table['O'] == 3);
        }
    }
}

File: Tic-Tac-Toe/CPP-TestProject/Judge.cpp
#include "Judge.h"
#include <vector>

using namespace std;


int Judge::CountMoves(vector<vector <char>> board) {

    int count = 0;

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {

            if (board[i][j] != ' ') {
                count++;
            }
        }
    }

    return count;
}


bool Judge::FindsWinner(vector<vector <char>> board) {

    for (int i = 0; i < 3; i++) {

        if (IsWinningRow(board, i)) {
            return true;
        }

        if (IsWinningColumn(board, i)) {
            return true;
        }
    }

    return HasWinningDiagonal(board);
}


bool Judge::IsWinningRow(vector<vector <char>> board, int i) {

    char team = board[i][0];

    if (team == ' ') {
        return false;
    }

    if (board[i][1] == team) {
        if (board[i][2] == team) {
            return true;
        }
    }

    return false;
}


bool Judge::IsWinningColumn(vector<vector <char>> board, int j) {

    char team = board[0][j];

    if (team == ' ') {
        return false;
    }

    if (board[1][j] == team) {
        if (board[2][j] == team) {
            return true;
        }
    }

    return false;
}


bool Judge::HasWinningDiagonal(vector<vector <char>> board) {

    char team = board[1][1];

    if (team == ' ') {
        return false;
    }

    if (board[0][0] == team) {
        if (board[2][2] == team) {
            return true;
        }
    }

    if (board[0][2] == team) {
        if (board[2][0] == team) {
            return true;
        }
    }

    return false;
}
